File: What's New in Location Technologies/TXLocationTechDemo/TXLocationTechOCDemo/TXLocationTechOCDemo/ViewController.m
//
//  ViewController.m
//  TXLocationTechDemo
//
//  Created by tingxins on 13/07/2017.
//  Copyright © 2017 tingxins. All rights reserved.
//

#import "ViewController.h"
#import <CoreLocation/CoreLocation.h>
#import <Contacts/Contacts.h>
#import <CoreMotion/CoreMotion.h>

@interface ViewController ()<CLLocationManagerDelegate>

@property (strong, nonatomic) CLLocationManager *manager;

@property (strong, nonatomic) CLGeocoder *geocoder;

@property (weak, nonatomic) IBOutlet UILabel *locateResultLabel;

@end

@implementation ViewController

- (CLGeocoder *)geocoder {
    if (_geocoder) return _geocoder;
    _geocoder = [[CLGeocoder alloc] init];
    return _geocoder;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    
    // 创建 Location Mannager
    [self setupLocationManager];
    
    // 程序进入后台通知
    [self addNotifications];
    
    // PostalAddress 解析测试
    [self postalAddressTest];
}

- (void)setupLocationManager {
    CLLocationManager *manager = [[CLLocationManager alloc] init];
    self.manager = manager;
    manager.allowsBackgroundLocationUpdates = YES;
    manager.delegate = self;
}

- (void)addNotifications {
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(startingBackgroundLocation) name:@"TXBackgroundStartUpdatingLocation" object:nil];
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(stoppingBackgroundLocation) name:@"TXBackgroundStopUpdatingLocation" object:nil];
}

#pragma mark - Customs

/** Notes: App has not access privacy-sensitive data without a usage description. The app's Info.plist must contain an NSLocationXXXXXDescription key with a string value explaining to the user how the app uses this data */
- (void)requestAuthorization {
    switch ([CLLocationManager authorizationStatus]) {
        case kCLAuthorizationStatusNotDetermined:
            [self.manager requestWhenInUseAuthorization];
            break;
        case kCLAuthorizationStatusAuthorizedAlways:
            [self.manager requestLocation];
            break;
        case kCLAuthorizationStatusAuthorizedWhenInUse:
            [self.manager requestAlwaysAuthorization];
            break;
        case kCLAuthorizationStatusRestricted:
            NSLog(@"Restricted");
            break;
        case kCLAuthorizationStatusDenied:
            NSLog(@"Denied");
            break;
            
        default:
            break;
    }
}

- (void)serialLocation:(CLLocation *)location {
    if (!location) return;
    NSLog(@"%s--%lf---%lf", __func__, location.coordinate.longitude,location.coordinate.latitude);
    [self.geocoder reverseGeocodeLocation:location completionHandler:^(NSArray<CLPlacemark *> * _Nullable placemarks, NSError * _Nullable error) {
        if (!error && placemarks.count) {
            NSLog(@"**********reverseGeocodeLocation************");
            CLPlacemark *placeMark = placemarks.lastObject;
            NSLog(@"%@", placeMark.addressDictionary);
            if (@available(iOS 11.0, *)) {
                NSLog(@"%@--%@", placeMark.name, placeMark.postalAddress);
            } else {
                // Fallback on earlier versions
                NSLog(@"%@", placeMark.name);
            }
            self.locateResultLabel.text = [NSString stringWithFormat:@"Current Location:%@", placeMark.name];
            NSLog(@"**********reverseGeocodeLocation************");
        }
    }];
}

- (void)postalAddressTest {
    CNMutablePostalAddress *homeAddress = [[CNMutablePostalAddress alloc] init];
    homeAddress.street = @"Keyan Road No.9 Nanshan";
    homeAddress.city = @"Shenzhen";
    homeAddress.state = @"Guangdong";
    homeAddress.postalCode = @"518000";
    homeAddress.country = @"China";
    homeAddress.ISOCountryCode = @"CN";
    
    if (@available(iOS 11.0, *)) {
        NSLocale *locale = [[NSLocale alloc] initWithLocaleIdentifier:@"zh_Hans_CN"];
        [self.geocoder geocodePostalAddress:homeAddress preferredLocale:locale completionHandler:^(NSArray<CLPlacemark *> * _Nullable placemarks, NSError * _Nullable error) {
            if (!error && placemarks.count) {
                NSLog(@"**********geocodePostalAddress-preferredLocale************");
                CLPlacemark *placemark = placemarks.lastObject;
                NSLog(@"%@--%@", placemark.name, placemark.postalAddress);
                NSLog(@"**********geocodePostalAddress-preferredLocale************");
            }
        }];
    } else {
        // Fallback on earlier versions
    }
}

- (IBAction)locating:(UIButton *)sender {
    NSLog(@"%s", __func__);
    [self.manager requestLocation];
}

#pragma mark - Background Notifications

- (void)startingBackgroundLocation {
    NSLog(@"%s", __func__);
    // 1.单次更新
//    [self.manager requestLocation];
    // 2.持续更新
    [self.manager startUpdatingLocation];
}

- (void)stoppingBackgroundLocation {
    NSLog(@"%s", __func__);
    [self.manager stopUpdatingLocation];
}

#pragma mark - CLLocationManagerDelegate

- (void)locationManager:(CLLocationManager *)manager didChangeAuthorizationStatus:(CLAuthorizationStatus)status {
    NSLog(@"RequestAuthorization:%d", status);
    [self requestAuthorization];
}

- (void)locationManager:(CLLocationManager *)manager didUpdateLocations:(NSArray<CLLocation *> *)locations {
    NSLog(@"%s", __func__);
    CLAuthorizationStatus status = [CLLocationManager authorizationStatus];
    if (status == kCLAuthorizationStatusAuthorizedWhenInUse ||
        status == kCLAuthorizationStatusAuthorizedAlways) {
        [self serialLocation:locations.lastObject];
    }
}

- (void)locationManager:(CLLocationManager *)manager didFailWithError:(NSError *)error {
    NSLog(@"didFailWithError:%@",error);
}

@end

File: What's New in Location Technologies/TXLocationTechDemo/TXLocationTechSwiftDemo/TXLocationTechSwiftDemo/ViewController.swift
//
//  ViewController.swift
//  TXLocationTechSwiftDemo
//
//  Created by tingxins on 05/08/2017.
//  Copyright © 2017 tingxins. All rights reserved.
//

import UIKit
import CoreLocation
import Contacts

class ViewController: UIViewController {
    
    @IBOutlet weak var locateResultLabel: UILabel!
    private lazy var manager: CLLocationManager = {
        let m = CLLocationManager()
        m.delegate = self
        return m
    }()
    
    private let geocoder: CLGeocoder = {
        return CLGeocoder()
    }()

    override func viewDidLoad() {
        super.viewDidLoad()
        
        setupLocationManager()
        
        addNotifications()
        
        postalAddressTest()
    }
    
    private func setupLocationManager() {
        manager.allowsBackgroundLocationUpdates = true
    }
}

// MARK: Custom
extension ViewController {
    
    /** Notes: App has not access privacy-sensitive data without a usage description. The app's Info.plist must contain an NSLocationXXXXXDescription key with a string value explaining to the user how the app uses this data */
    private func requestAuthorization() {
        switch CLLocationManager.authorizationStatus() {
        case .notDetermined:
            manager.requestWhenInUseAuthorization()
        case .authorizedAlways:
            manager.requestLocation()
        case .authorizedWhenInUse:
            manager.requestAlwaysAuthorization()
        case .restricted:
            print("Restricted")
        case .denied:
            print("Denied")
        }
    }
    
    private func serial(with location: CLLocation?) {
        guard let loc = location else { return }
        print("longitude:\(loc.coordinate.longitude), latitude:\(loc.coordinate.latitude)")
        geocoder.reverseGeocodeLocation(loc) { (placemarks, error) in
            
            if (error != nil) { return }
            
            guard let pms = placemarks else { return }
            
            let placemark = pms.last!
            if #available(iOS 11.0, *) {
                print("\(placemark.name ?? "none")--\(placemark.postalAddress ?? CNPostalAddress())")
            }else {
                print("\(placemark.name ?? "none")")
            }
            
            self.locateResultLabel.text = placemark.name
        }
    }
    
    private func postalAddressTest() {
        let postalAddress = CNMutablePostalAddress()
        postalAddress.street = "Keyan Road No.9 Nanshan"
        postalAddress.city = "Shenzhen"
        postalAddress.state = "Guangdong"
        postalAddress.postalCode = "518000"
        postalAddress.country = "China"
        postalAddress.isoCountryCode = "CN"
        
        if #available(iOS 11.0, *) {
            let locale = Locale.init(identifier: "zh_Hans_CN")
            geocoder.geocodePostalAddress(postalAddress, preferredLocale: locale, completionHandler: { (placemarks, error) in
                
                if (error != nil) { return }
                
                guard let pms = placemarks else { return }
                
                let placemark = pms.last!
                
                print("\(placemark.name ?? "none")--\(placemark.postalAddress ?? CNPostalAddress())")
            })
        }
    }
}

// MARK: Background Notifications
extension ViewController {
    
    @objc func startingBackgroundLocation() {
        manager.startUpdatingLocation()
    }
    
    @objc func stoppingBackgroundLocation() {
        manager.stopUpdatingLocation()
    }
    
    private func addNotifications() {
        NotificationCenter.default.addObserver(self, selector: #selector(startingBackgroundLocation), name: NSNotification.Name("BackgroundStartUpdatingLocation"), object: nil)
        NotificationCenter.default.addObserver(self, selector: #selector(startingBackgroundLocation), name: NSNotification.Name("BackgroundStopUpdatingLocation"), object: nil)
    }
}

// Target
extension ViewController {
    @IBAction func locating(_ sender: UIButton) {
        print(#function)
        manager.requestLocation()
    }
}

// MARK: CLLocationManagerDelegate
extension ViewController: CLLocationManagerDelegate {
    func locationManager(_ manager: CLLocationManager, didChangeAuthorization status: CLAuthorizationStatus) {
        print(#function)
        requestAuthorization()
    }
    
    func locationManager(_ manager: CLLocationManager, didUpdateLocations locations: [CLLocation]) {
        print(#function)
        let status = CLLocationManager.authorizationStatus()
        if status == .authorizedAlways ||
           status == .authorizedWhenInUse {
            serial(with: locations.last)
        }
    }
    
    func locationManager(_ manager: CLLocationManager, didFailWithError error: Error) {
        print(#function)
    }
}
